File: privacy results.m
%% privacy_results.m
% MATLAB script version of privacy_results.py
clear; clc; tic;

% Epsilon values
eps_vals = [0.01, 0.1, 1.0, 10, 20];
ne = numel(eps_vals);

% Preallocate
FM_noisy2     = zeros(1, ne);
FM_noiseless  = zeros(1, ne);
dpsgd2        = zeros(1, ne);
dpsgd3        = zeros(1, ne);
nonprivate    = zeros(1, ne);
nonprivate_time = zeros(1, ne);
FM_time       = zeros(1, ne);
dpsgd_time    = zeros(1, ne);
PALM_noisy2   = zeros(1, ne);
PALM_noiseless= zeros(1, ne);
PALM_time     = zeros(1, ne);

% --- Load FM accuracy ---
for i = 1:ne
    FM_noiseless(i) = read_mean(sprintf('results/FMaccuracy_noislessInp_%d.txt', i-1));
    FM_noisy2(i)    = read_mean(sprintf('results/FMaccuracy_noisyInp_%d_1_5.txt', i-1));
end

% --- Load DPSGD accuracy ---
for i = 1:ne
    dpsgd2(i) = read_mean(sprintf('results/dpsgdaccuracy_%d_1_0_1.txt', i-1));
    dpsgd3(i) = read_mean(sprintf('results/dpsgdaccuracy_%d_1_1_5.txt', i-1));
end

% --- Load Non-private ---
for i = 1:ne
    nonprivate(i) = read_mean('results/nonprivate_0.txt');
end

% --- Time data ---
for i = 1:ne
    nonprivate_time(i) = read_mean(sprintf('results/nonPrivate_time_%d_0_0_1.txt', i-1));
    FM_time(i)         = read_mean(sprintf('results/fm_time_%d_1_0_1.txt', i-1));
    dpsgd_time(i)      = read_mean(sprintf('results/dpsgd_time_%d_1_0_1.txt', i-1));
end

% --- PALM ---
for i = 1:ne
    PALM_noiseless(i) = read_mean(sprintf('results/PALMaccuracy_noislessInp_%d.txt', i-1));
    PALM_noisy2(i)    = read_mean(sprintf('results/PALMaccuracy_noisyInp_%d_1_5.txt', i-1));
    PALM_time(i)      = read_mean(sprintf('results/PALM_time_%d_1_0_1.txt', i-1));
end

% --- Debug prints ---
diff1 = (mean(PALM_noiseless, 2) - mean(dpsgd2, 2)) ./ mean(dpsgd2, 2) * 100;
fprintf('diff1: %.6f\n', diff1);
diff2 = (mean(PALM_noisy2, 2) - mean(dpsgd3, 2)) ./ mean(dpsgd3, 2) * 100;
fprintf('diff2: %.6f\n', diff2);
fprintf('dpsgd_time: %s\n', mat2str(dpsgd_time, 6));
fprintf('PALM_time: %s\n', mat2str(PALM_time, 6));
time_diff1 = (mean(dpsgd_time, 2) - mean(PALM_time, 2)) ./ mean(dpsgd_time, 2) * 100;
fprintf('time_diff1: %.6f\n', time_diff1);

% --- Plot ---
figure('Units','pixels','Position',[100 100 800 600]); hold on;
xidx = 1:ne;
plot(xidx, nonprivate, '-.o', 'LineWidth', 3, 'MarkerSize', 10);
plot(xidx, FM_noiseless, '-.*', 'LineWidth', 3, 'MarkerSize', 10);
plot(xidx, FM_noisy2, '-.d', 'LineWidth', 3, 'MarkerSize', 10);
plot(xidx, dpsgd2, '-.s', 'LineWidth', 3, 'MarkerSize', 10);
plot(xidx, dpsgd3, '-.x', 'LineWidth', 3, 'MarkerSize', 10);
plot(xidx, PALM_noiseless, '--^', 'LineWidth', 3, 'MarkerSize', 10);
plot(xidx, PALM_noisy2, '--h', 'LineWidth', 3, 'MarkerSize', 10);
hold off; ylim([0.37 1]); grid on;

ax = gca;
ax.XTick = xidx;
ax.XTickLabel = arrayfun(@(v) num2str(v), eps_vals, 'UniformOutput', false);
set(ax, 'XDir', 'reverse');

ylabel('accuracy', 'FontSize', 16, 'Interpreter','none');
xlabel('privacy budget $\epsilon$', 'FontSize', 16, 'Interpreter','latex');

legend(...
    {'non-private (standard BP)', ...
     'FM noiseless inputs (BK)', ...
     'FM noisy inputs, $\sigma=5$ (BK)', ...
     'DP-SGD noiseless inputs (BK+GC)', ...
     'DP-SGD noisy inputs, $\sigma=5$ (BK+GC)', ...
     'SPOF noiseless inputs (BK)', ...
     'SPOF noisy inputs, $\sigma=5$ (BK)'}, ...
     'Interpreter','latex', ...
     'Location','southwest');

if ~exist('results', 'dir'); mkdir('results'); end
print(gcf, fullfile('results','results'), '-dpdf', '-fillpage');

fprintf('Done in %.2f seconds\n', toc);

%% Helper function
function m = read_mean(fname)
    if ~isfile(fname)
        error('File not found: %s', fname);
    end
    data = readmatrix(fname);
    data = data(~isnan(data));
    if isempty(data)
        m = NaN;
    else
        m = mean(data(:));
    end
end

File: privacy_results.py
import os
os.system('clear')  

import matplotlib
matplotlib.use("Agg")  

import numpy as np
import pandas as pd
import time
import math
from numpy import array
import matplotlib.pyplot as plt
import matplotlib.font_manager as font_manager
from matplotlib.ticker import MultipleLocator
plt.rcParams.update({
    "text.usetex": True,
    "font.family": "Computer Modern"
})
plt.rcParams.update({'font.size': 14})
start_time = time.time()
from scipy.interpolate import make_interp_spline

#eps=[0.001, 0.01, 0.1, 1.0, 10.0] #use the results2 folder
eps=[0.1, 0.2, 0.4, 0.8, 1.6]
FM_noisy=np.zeros(len(eps))
FM_noisy2=np.zeros(len(eps))
FM_noiseless=np.zeros(len(eps))
dpsgd=np.zeros(len(eps))
dpsgd2=np.zeros(len(eps))
dpsgd3=np.zeros(len(eps))
nonprivate=np.zeros(len(eps))
nonprivate_time=np.zeros(len(eps))
FM_time=np.zeros(len(eps))
dpsgd_time=np.zeros(len(eps))

# --- PALM arrays ---
PALM_noisy=np.zeros(len(eps))
PALM_noisy2=np.zeros(len(eps))
PALM_noiseless=np.zeros(len(eps))
PALM_time=np.zeros(len(eps))

# --- Load FM accuracy ---
for i in range(len(eps)):
    accuracy=np.loadtxt(f"results_2/FMaccuracy_noislessInp_{i}.txt") 
    FM_noiseless[i]=array(accuracy).reshape(1, -1).mean(1)

    #accuracy=np.loadtxt(f"results_2/FMaccuracy_noisyInp_{i}_1_1.txt") 
    #FM_noisy[i]=array(accuracy).reshape(1, -1).mean(1)

    accuracy=np.loadtxt(f"results_2/FMaccuracy_noisyInp_{i}_1_5.txt") 
    FM_noisy2[i]=array(accuracy).reshape(1, -1).mean(1)

# --- Load DPSGD accuracy ---
for i in range(len(eps)):
    accuracy=np.loadtxt(f"results_2/dpsgdaccuracy_{i}_1_0_1.txt") 
    dpsgd2[i]=array(accuracy).reshape(1, -1).mean(1)

    accuracy=np.loadtxt(f"results_2/dpsgdaccuracy_{i}_1_1_5.txt") 
    dpsgd3[i]=array(accuracy).reshape(1, -1).mean(1) 

# --- Load Non-private ---
for i in range(len(eps)):
    accuracy=np.loadtxt("results_2/nonprivate_0.txt") 
    nonprivate[i]=array(accuracy).reshape(1, -1).mean(1)

# --- Load time: non-private ---
for i in range(1):
    time_=np.loadtxt(f"results_2/nonPrivate_time_{i}_1_1_5.txt") 
    nonprivate_time[i]=array(time_).reshape(1, -1).mean(1)

# --- Load time: FM ---
for i in range(len(eps)):
    time_=np.loadtxt(f"results_2/FM_time_{i}_1_0_5.txt") 
    FM_time[i]=array(time_).reshape(1, -1).mean(1)

# --- Load time: DPSGD ---
for i in range(len(eps)):
    time_=np.loadtxt(f"results_2/dpsgd_time_{i}_1_0_1.txt") 
    dpsgd_time[i]=array(time_).reshape(1, -1).mean(1)

# --- Load PALM results ---
for i in range(len(eps)):
    accuracy=np.loadtxt(f"results_2/PALMaccuracy_noislessInp_{i}.txt")
    PALM_noiseless[i]=array(accuracy).reshape(1, -1).mean(1)

    #accuracy=np.loadtxt(f"results_2/PALMaccuracy_noisyInp_{i}_1_1.txt")
    #PALM_noisy[i]=array(accuracy).reshape(1, -1).mean(1)

    accuracy=np.loadtxt(f"results_2/PALMaccuracy_noisyInp_{i}_1_5.txt")
    PALM_noisy2[i]=array(accuracy).reshape(1, -1).mean(1)

    time_=np.loadtxt(f"results_2/PALM_time_{i}_1_0_5.txt")
    PALM_time[i]=array(time_).reshape(1, -1).mean(1)

# --- Optional debug prints ---
diff1=(PALM_noiseless.mean(0)-dpsgd2.mean(0))/dpsgd2.mean(0)*100
print('diff1: ',diff1)
diff2=(PALM_noisy2.mean(0)-dpsgd3.mean(0))/dpsgd3.mean(0)*100
print('diff2: ',diff2)
#diff2=(dpsgd_time.mean(0)-PALM_time.mean(0))*100
print('dpsgd_time: ',dpsgd_time)
print('PALM_time: ',PALM_time)
time_diff1=(dpsgd_time.mean(0)-PALM_time.mean(0))/dpsgd_time.mean(0)*100
print('time_diff1: ',time_diff1)

# --- Plot ---
fig1 = plt.figure(figsize=(8,6))
x_indices = range(len(eps))

plt.plot(
    x_indices, nonprivate, '-.o',
    x_indices, FM_noiseless, '-.*',
    #x_indices, FM_noisy, '-.<',
    x_indices, FM_noisy2, '-.d',
    x_indices, dpsgd2, '-.s',
    x_indices, dpsgd3, '-.x',
    x_indices, PALM_noiseless, '--^',
    #x_indices, PALM_noisy, '--v',
    x_indices, PALM_noisy2, '--h',
    linewidth=3, markersize=10
)

plt.xticks(x_indices, eps)
plt.ylim([0.7, 1])
plt.gca().invert_xaxis()
plt.grid(True, which="both")
plt.ylabel('accuracy', fontsize=16)
plt.xlabel('privacy budget $\epsilon$', fontsize=16)
plt.gca().legend((
    'non-private',
    'FM noiseless inputs',
    #'FM noisy inputs, $\sigma=1$',
    'FM noisy inputs, $\sigma=5$',
    'DP-SGD noiseless inputs',
    'DP-SGD noisy inputs, $\sigma=5$',
    'SPOF noiseless inputs',
    #'PALM noisy inputs, $\sigma=1$',
    'SPOF noisy inputs, $\sigma=5$'
), loc="center", bbox_to_anchor=(0.325, 0.2), borderaxespad=0., labelspacing=0.25)

fig1.savefig('results.pdf', bbox_inches='tight')

# Average ONLY times
FM_time_avg = FM_time
DPSGD_time_avg = dpsgd_time
PALM_time_avg = PALM_time
NonPrivate_time_avg = nonprivate_time

# Build table
table = np.vstack([
    eps,
    FM_time_avg,
    DPSGD_time_avg,
    PALM_time_avg,
    NonPrivate_time_avg
])

df = pd.DataFrame(
    table,
    index=["epsilon", "FM", "DPSGD", "PALM", "NonPrivate"]
)
